pkg/operator/ceph/object/bucket/rgw-handlers.go

```diff
@@ -62,33 +109,52 @@ func (p *Provisioner) genUserName(obc *v1alpha1.ObjectBucketClaim) string {
-	if len(bucketName) > 0 {
-		// delete bucket with purge option to remove all objects
-		thePurge := true
-		err := p.adminOpsClient.RemoveBucket(p.clusterInfo.Context, admin.Bucket{Bucket: bucketName, PurgeObject: &thePurge})
-		if err == nil {
-			logger.Infof("bucket %q successfully deleted", bucketName)
-		} else if errors.Is(err, admin.ErrNoSuchBucket) {
-			// opinion: "not found" is not an error
-			logger.Infof("bucket %q does not exist", bucketName)
-		} else if errors.Is(err, admin.ErrNoSuchKey) {
-			// ceph might return NoSuchKey than NoSuchBucket when the target bucket does not exist.
-			// then we can use GetBucketInfo() to judge the existence of the bucket.
-			// see: https://github.com/ceph/ceph/pull/44413
-			_, err2 := p.adminOpsClient.GetBucketInfo(p.clusterInfo.Context, admin.Bucket{Bucket: bucketName, PurgeObject: &thePurge})
-			if errors.Is(err2, admin.ErrNoSuchBucket) {
-				logger.Infof("bucket info %q does not exist", bucketName)
-			} else {
-				return errors.Wrapf(err, "failed to delete bucket %q (could not get bucket info)", bucketName)
-			}
+func (p *Provisioner) deleteBucket(bucketName string) error {
+	logger.Infof("deleting Ceph bucket %q", bucketName)
+	// delete bucket with purge option to remove all objects
+	thePurge := true
+	err := p.adminOpsClient.RemoveBucket(p.clusterInfo.Context, admin.Bucket{Bucket: bucketName, PurgeObject: &thePurge})
+	if err == nil {
+		logger.Infof("bucket %q successfully deleted", bucketName)
+	} else if errors.Is(err, admin.ErrNoSuchBucket) {
+		// opinion: "not found" is not an error
+		logger.Infof("bucket %q does not exist", bucketName)
+	} else if errors.Is(err, admin.ErrNoSuchKey) {
+		// ceph might return NoSuchKey than NoSuchBucket when the target bucket does not exist.
+		// then we can use GetBucketInfo() to judge the existence of the bucket.
+		// see: https://github.com/ceph/ceph/pull/44413
+		_, err2 := p.adminOpsClient.GetBucketInfo(p.clusterInfo.Context, admin.Bucket{Bucket: bucketName, PurgeObject: &thePurge})
+		if errors.Is(err2, admin.ErrNoSuchBucket) {
+			logger.Infof("bucket info %q does not exist", bucketName)
 		} else {
-			return errors.Wrapf(err, "failed to delete bucket %q", bucketName)
+			return errors.Wrapf(err, "failed to delete bucket %q (could not get bucket info)", bucketName)
 		}
+	} else {
+		return errors.Wrapf(err, "failed to delete bucket %q", bucketName)
+	}
+
+	return nil
+}
+
+// Delete the user *if it was created by OBC*. Will not delete externally
+// managed users / users not created by obc.
+func (p *Provisioner) deleteOBUser(ob *bktv1alpha1.ObjectBucket) error {
+	// the OBC is needed to determine if the user was created by the provisioner
+	obc := &bktv1alpha1.ObjectBucketClaim{}
+	namespacedName := types.NamespacedName{
+		Name:      ob.Spec.ClaimRef.Name,
+		Namespace: ob.Spec.ClaimRef.Namespace,
 	}
-	if len(p.cephUserName) > 0 {
+	err := p.context.Client.Get(context.Background(), namespacedName, obc)
```

Comment:
Is the lib-bucket clientset already setup some some place that is accessible from the provisioner?

Reply:
I'm not sure I follow the question. Are you asking if it's kosher to get the OBC here from the bucket handler? If so, I would say the answer is generally that we should avoid using OB/OBC info that isn't passed to us -- but I think we might already do that someplace already? It's hard because l-b-p isn't really well-implemented

Reply:
That wasn't what I was asking.  Retrieving the obc here is the only way to address [this request](https://github.com/rook/rook/pull/15311#discussion_r1937770683) to check the `bucketOwner`  as `Delete()` is only called with the ob.

The question was is if the statically typed lib-bucket clientset is accessible from somewhere instead of using the dynamic k8s client.  It isn't part of `clusterd.Context`: https://github.com/rook/rook/blob/master/pkg/clusterd/context.go#L31

Reply:
I think the right pre-existing mechanism for this is to start adding `bucketOwner` to the OB's additional state info. This is where we keep info about the Ceph object store name and namespace. 

Here's where Rook does some ns/name work:

https://github.com/rook/rook/blob/a55fdb3fbeabe0155272ea048fc19d1b4c6e9209/pkg/operator/ceph/object/bucket/util.go#L158-L168

Here's the PR where Yuval used this mechanism for work on the notifications feature: https://github.com/rook/rook/pull/9150

Reply:
I didn't know that changing the ob was fair game. I'll rework the logic.

Reply:
I've added a `bucketOwner` key to the ob as well.